### DISCUSSION TITLE
Postgres enum type not working with  `select_only().column()`

### DISCUSSION THREAD

**Discussion by MaderNoob** [SeaQL/sea-orm#843] [Q&A] [answered]
## Description

When using a postgres enum as the type of a column, fetching normally using `Entity::find()` works, but using `.select_only().column()` with that column doesn't work, and gives the following error:

error occurred while decoding column "<column name>": mismatched types; Rust type `core::option::Option<alloc::string::String>` (as SQL type `TEXT`) is not compatible with SQL type `<enum type name>`

## Steps to Reproduce

1. define an enum and implement `Iden` for it: 
```rust
#[derive(Debug, EnumIter, Display)]
#[strum(serialize_all = "snake_case")]
pub enum SomeKind {
    #[strum(serialize = "some_kind")]
    Type,

    A,B,C
}
impl Iden for SomeKind {
    fn unquoted(&self, s: &mut dyn std::fmt::Write) {
        write!(s, "{}", self).unwrap()
    }
}
```
2. create a postgres type for it: 
```rust
        manager
            .create_type(
                sea_query::extension::postgres::Type::create()
                    .as_enum(SomeKind::Type)
                    .values(SomeKind::iter())
                    .to_owned(),
            )
            .await?;
```
4. create a table which uses the type in one of its columns:
```rust
#[derive(Iden)]
pub enum SomeTable {
    Table,
    Id,
    Kind,
}

        manager
            .create_table(
                sea_query::Table::create()
                    .table(SomeTable::Table)
                    .col(
                        ColumnDef::new(SomeTable::Id)
                            .integer()
                            .auto_increment()
                            .primary_key()
                            .not_null(),
                    )
                    .col(
                        ColumnDef::new(SomeTable::Kind)
                            .text()
                            .enumeration(format!("{}", SomeKind::Type), SomeKind::iter())
                            .not_null(),

                    )
                    .to_owned(),
            )
            .await?;

```
5. use the codegen: `sea-orm-cli generate -o ...`
6. try to fetch a record using `select_only().column()`:
```rust
struct SomeTableRecordKind{
    kind: entity::sea_orm_active_enums::SomeKind,
} 

let db = ...;
entity::some_table::Entity::find().select_only().column(some_table::Column::Kind).into_model::<SomeTableRecordKind>().one(db).await.unwrap();
```

### Expected Behavior

The query should work properly

### Actual Behavior

An error is returned:

error occurred while decoding column "<column name>": mismatched types; Rust type `core::option::Option<alloc::string::String>` (as SQL type `TEXT`) is not compatible with SQL type `<enum type name>`

### Reproduces How Often

always

## Versions

│   └── sea-orm v0.7.1
│       ├── sea-orm-macros v0.7.0 (proc-macro)
│       ├── sea-query v0.23.0
│       │   ├── sea-query-derive v0.2.0 (proc-macro)
│       ├── sea-strum v0.23.0
│       │   └── sea-strum_macros v0.23.0 (proc-macro)
│   ├── sea-schema v0.7.1
│   │   ├── sea-orm v0.7.1 (*)
│   │   ├── sea-query v0.22.0
│   │   │   └── sea-query-derive v0.2.0 (proc-macro) (*)
│   │   ├── sea-schema-derive v0.1.0 (proc-macro)
├── sea-orm v0.7.1 (*)
├── sea-orm-rocket v0.5.0
│   └── sea-orm-rocket-codegen v0.5.0 (proc-macro)

`uname -a`:
Linux arch 5.17.5-arch1-1 #1 SMP PREEMPT Wed, 27 Apr 2022 20:56:11 +0000 x86_64 GNU/Linux

`pacman -Q postgresql`:
postgresql 14.2-1

## Additional Information

I compared queries without `.select_only().column()` and with, and noticed that when not using `.select_only().column()`, a cast is performed on the column: `CAST("some_table"."kind" AS text)`, but when using `.select_only().column()` no cast is performed.

**Comment by billy1624** [comment, marked as answer]
Hey @MaderNoob, sorry for the delay. You have to cast the enum explicitly. For example,
https://github.com/SeaQL/sea-orm/blob/4301383b409737f85378a1da8a4c122478248ba0/src/query/select.rs#L124-L134

**Reply by encody** [reply]
I'm pretty confused by this. I used the CLI to codegen my entities. The database I'm interacting with uses a lot of enums&mdash;do I need to write something like this for every column that uses an enum? If so, why doesn't the CLI / macros write this for me?

**Reply by encody** [reply]
Sorry, that was probably not a very answerable question. I was able to get it working using the following wrapper function:

```rust
pub fn col_enum<T>(column: T) -> SimpleExpr
where
    T: IntoColumnRef,
{
    Expr::col(column).as_enum(Alias::new("text"))
}
```

Later called like so:

```rust
MyTable::find()
        .column_as(col_enum(my_table::Column::MyColumn), "my_column")
```

Just wondering:
1. Why this is necessary.
2. If this could be documented a little more clearly. Took me a couple hours to figure out why my query broke.

**Comment by tyt2y3** [comment]
Looks like we are doing some magic inside `fn column_list(&self) -> Vec<SimpleExpr> {` of `select.rs`.
It should also be possible to do the same in `fn column()` ? @billy1624 
Then it will be transparent to the user?

**Comment by billy1624** [comment]
For those of you who bump into this thread, there is a tracking issue for you:
- https://github.com/SeaQL/sea-orm/issues/960
